syntax = "proto3";

package vnpt.audio.asr;

option cc_enable_arenas = true;

import 'vnpt_audio.proto';

service VnptSpeechRecognition {
    rpc Recognize(RecognizeRequest) returns (RecognizeResponse) {}
    rpc StreamingRecognize(stream StreamingRecognizeRequest) returns (stream StreamingRecognizeResponse) {}
}

message RecognizeRequest {
    RecognitionConfig config = 1;
    bytes audio = 2;
}

message StreamingRecognizeRequest {
    oneof streaming_request {
        StreamingRecognitionConfig streaming_config = 1;
        bytes audio_content = 2;
    }
}

message RecognitionConfig {
    AudioEncoding encoding = 1;
    int32 sample_rate_hertz = 2;
    string language_code = 3;
    int32 max_alternatives = 4;
    repeated SpeechContext speech_contexts = 6;
  	int32 audio_channel_count = 7;
    bool enable_word_time_offsets = 8;
  	bool enable_automatic_punctuation = 11;
  	bool enable_separate_recognition_per_channel = 12;
  	string model = 13;
    bool verbatim_transcripts = 14;
    map<string, string> custom_configuration = 24;
}

message StreamingRecognitionConfig {
    RecognitionConfig config = 1;
    bool interim_results = 2;
}

message SpeechContext {
  repeated string phrases = 1;
  float boost = 4;
}

message RecognizeResponse {
    repeated SpeechRecognitionResult results = 1;
}

message SpeechRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
  int32 channel_tag = 2;
  float audio_processed = 3;
}
message SpeechRecognitionAlternative {
  string transcript = 1;
  float confidence = 2;
  repeated WordInfo words = 3;
}

message WordInfo {
  int32 start_time = 1;
  int32 end_time = 2;
  string word = 3;
}

message StreamingRecognizeResponse {
    repeated StreamingRecognitionResult results = 1;
}

message StreamingRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
  bool is_final = 2;
  float stability = 3;
  int32 channel_tag = 5;
  float audio_processed = 6;
}
